components/playlist/browser/playlist_tab_helper.cc

```diff
@@ -351,6 +399,35 @@ void PlaylistTabHelper::OnFoundMediaFromContents(
   for (auto& observer : observers_) {
     observer.OnFoundItemsChanged(found_items_);
   }
+
+  if (found_items_.size() &&
+      !service_->ShouldExtractMediaFromBackgroundWebContents(found_items_)) {
+    // Wait until we find media or timeout. Some pages might not have media
+    // initially but update media later.
+    auto callbacks =
+        std::move(media_extracted_from_background_web_contents_callbacks_);
+    for (auto& callback : callbacks) {
+      std::move(callback).Run(true);
+    }
+    media_extraction_from_background_web_contents_timer_.Stop();
```

Comment:
Why is the tab helper responsible for this? Shouldn't the service handle the timeout and actually cancel the request instead of returning it later when it might possibly conflict with another request that has started since then? This entire process appears to have a number of potential race conditions and I don't think the various flags used in this class are reliable across multiple requests. Putting all this logic in the tab helper also means that it cannot be shared with ios. I would suggest a cancellable task tracker and/or generating a new UnguessableToken in PrimaryPageChanged